src/sctools/barcode.py

```diff
@@ -0,0 +1,134 @@
+import itertools
+from collections import Counter, Mapping
+import numpy as np
+from .encodings import TwoBit
+from .stats import base4_entropy
+
+
+class BarcodeBase:
+    """Abstract base class for sets of barcodes, encoded in 2bit representation for fast operations
+
+    :method summarize_hamming_distances:  returns descriptive statistics on hamming distances
+      between pairs of barcodes
+
+    """
+
+    def __init__(self, barcode_set, barcode_length):
+        """
+        :param Counter barcode_set: dictionary
+        :param int barcode_length: the length of all barcodes in the set. Different-length barcodes
+          are not supported
+        """
+        if not isinstance(barcode_set, Mapping):
+            raise TypeError('barcode set must be a dict-like object mapping barcodes to counts')
+        self._data = barcode_set
+        self._barcode_length = barcode_length
+
+    def __contains__(self, item):
+        return item in self._data
+
+    def __iter__(self):
+        return iter(self._data)
+
+    def __len__(self):
+        return len(self._data)
+
+    def __getitem__(self, item):
+        return self._data[item]
+
+    def summarize_hamming_distances(self):
+        """returns descriptive statistics on hamming distances between pairs of barcodes"""
+        distances = []
+        for a, b in itertools.combinations(self, 2):
+            distances.append(bin(a ^ b).count('1'))
+        distances.sort()
+
+        return {
+            'minimum': distances[0],
+            '25th percentile': distances[int(len(distances) * .25)],
+            'median': distances[int(len(distances) * .5)],
+            'average': sum(distances) / len(distances),
+            '75th percentile': distances[int(len(distances) * .75)],
+            'maximum': distances[-1]
+        }
+
+    def base_frequency(self, weighted=False):
+        """return the frequency of each base at each position in the barcode set
+
+        :param bool weighted: (Default False) if True, each barcode is counted once for each time it
+          was observed
+        :return np.array: barcode_length x 4 2d array
+        """
+        base_counts_by_position = np.zeros((self._barcode_length, 4), dtype=np.uint64)
+        keys = np.fromiter(self._data.keys(), dtype=np.uint64)
+        values = np.fromiter(self._data.values(), dtype=np.uint64)
+
+        for i in reversed(range(self._barcode_length)):
+            binary_base_representations, counts = np.unique(keys & 3, return_counts=True)
+            if weighted:  # todo weighted not working, values multiplication does not work
+                base_counts_by_position[i, binary_base_representations] = counts  # * values
```

Comment:
Maybe this should throw a NotImplementedError, or else take out this option altogether for now.

Reply:
it should throw a NotImplementedError or I can extract it to another open branch and add an issue. I think this will be useful, but we can drop it from this upload as I haven't figured out how to do this efficiently, yet.